Add internal memory to CFG

efferifick opened the pull request:
This commit adds an internal memory region to the CFG.
By default, the internal memory region is created from the heap
region. CFG edges are now allocated to this internal memory
region.

Signed-off-by: Erick Ochoa <eochoa@ualberta.ca>

Needed for this [openj9 issue](https://github.com/eclipse/openj9/issues/6099)
Will be needed for this [openj9 PR](https://github.com/eclipse/openj9/pull/6141)

andrewcraik commented on compiler/infra/Cfg.hpp:
```diff
@@ -34,8 +34,8 @@ class CFG : public OMR::CFGConnector
 
-   CFG(TR::Compilation *comp, TR::ResolvedMethodSymbol *method) :
-      OMR::CFGConnector(comp, method) {}
+   CFG(TR::Compilation *comp, TR::ResolvedMethodSymbol *method, TR::Region *region = NULL) :
```
Rather than adding a `TR::Region *` parameter why not just create another constructor overload that takes the region as a reference. The version without a region will default to the heap region and the other will default to the region supplied. There is just a coding convention in the compiler that we pass regions by reference precisely because NULL regions make no sense...

efferifick commented:
@andrewcraik I removed the region pointer constructor and added the region reference constructor. I also added the optional `TR_AllocationKind` parameters used in `addEdge` and `addExceptionEdge` to allow master openj9 to compile.

andrewcraik commented:
Is the plan to remove the `addEdge` and `addExceptionEdge` versions taking the `TR_AllocationKind` once OpenJ9 stops using them?

efferifick commented:
@andrewcraik yes. The only place where `TR_AllocationKind` is being used is in `TR_J9EstimateCodeSize::realEstimateSize`. The [openj9 commit](https://github.com/eclipse/openj9/pull/6141) removes them. After removing them from OpenJ9, then we can remove the optional parameters from OMR without worrying about building issues.

andrewcraik commented:
Great - second point - I'm thinking that moving the field initialization into an init function called from the constructor rather than duplicating all of the initial values to reduce the need to double edit in the future if we need to change default values. Thoughts?

efferifick commented:
Yes, that sounds good. I'll make the changes.

andrewcraik commented:
Can you also mark the allocationKind APIs deprecated since they will be in limbo while OpenJ9 catches up so that the APIs can disappear completely...

efferifick commented:
Wait, doesn't `_internalRegion` need to be in the initialization list? The constructor for `TR::Region` takes an argument and therefore deferring the initialization to even inside the constructor would result in a build error, wouldn't it?

andrewcraik commented:
so you could only do _internalRegion or other fields that need direct initialization in the constructor and move everything else out - that I think would leave only the regions - one of which will vary depending on the region being supplied or not...

efferifick commented:
I added the previous implementation of `addEdge` and `addExceptionEdge` because without them `TR_J9EstimateCodeSize::realEstimateSize` will change semantics slightly. Instead of allocating in the stack region, the CFGs will be allocated in the heap region. After this PR is merged, my intention is to swith `TR_J9EstimateCodeSize::realEstimateSize` to use the new API and submit another PR here to remove the deprecated API since it is only used in `TR_J9EstimateCodeSize::realEstimateSize`.

andrewcraik commented:
I think this is looking pretty good - @Leonardo2718 would you mind taking a look?

Leonardo2718 commented on compiler/infra/OMRCfg.cpp:
```diff
@@ -228,7 +259,70 @@ OMR::CFG::addExceptionEdge(
+      // The existing catch block is going to be considered first.  Don't add an edge to the new catch
+      // block if the existing one catches everything that the new one catches.
+      //
+      /////void * newEC = newCatchBlock->getExceptionClass(), * existingEC = existingCatchBlock->getExceptionClass();
```
Does this commented out code serve any purpose?

efferifick replied:
The commented out code is also found in the deprecated API. This is just a literal copy paste. By this I mean: since the code was commented out originally by someone else, and it has been commented since the [initial contribution](https://github.com/eclipse/omr/blame/master/compiler/infra/OMRCfg.cpp#L202), I don't believe it serves any purpose. I can delete the comments if you want.

Regarding the code duplication, I don't think it should be an issue. As soon as this change (and one in OpenJ9) are pulled, then I can come back and delete the previous implementation.

Leonardo2718 replied:
I see. I'm fine with keeping the comments if they are just copy-pasted from old code. I'll leave it up to you to decide if you want to remove them or open an issue about it instead (it would be a good beginner item).

Leonardo2718 commented on compiler/infra/OMRCfg.cpp:
```diff
@@ -228,7 +259,70 @@ OMR::CFG::addExceptionEdge(
+      /////void * newEC = newCatchBlock->getExceptionClass(), * existingEC = existingCatchBlock->getExceptionClass();
 
+      if (existingCatchBlock->getCatchType() == 0 ||
+          /////(newEC && existingEC && isInstanceOf(newEC, existingEC)) ||
```
Does this commented out code serve any purpose?

efferifick replied:
Same for this code.

andrewcraik commented:
@genie-omr build all

andrewcraik commented:
@genie-omr build all

fjeremic commented:
@genie-omr build all

fjeremic commented:
@andrewcraik I think the reason the build triggers didn't work for you is because you were not yet added to the admin list on the Jenkins side. I've added your GitHub username now. I think you should be able to trigger builds. Give it a shot here perhaps.

andrewcraik commented on compiler/infra/OMRCfg.hpp:
```diff
@@ -161,8 +174,27 @@ class CFG
+    *     addEdge(TR::CFGNode *f, TR::CFGNode *t);
+    */
+   TR::CFGEdge *addEdge(TR::CFGNode *f, TR::CFGNode *t, TR_AllocationKind heapAlloc);
+   TR::CFGEdge *addEdge(TR::CFGNode *f, TR::CFGNode *t);
```
Can you please provide doxygen for the non-deprecated APIs which people should be using long term?

efferifick replied:
Will do.

andrewcraik commented:
Overall I think this is pretty much ready - just a bit of doxygen missing. @0xdaryl and @mstoodle you are listed as pending reviewers? Could you have a look or remove the review request?

andrewcraik commented:
@genie-omr build all

andrewcraik commented:
@genie-omr build all

andrewcraik commented on compiler/infra/OMRCfg.hpp:
```diff
@@ -97,34 +97,47 @@ class CFG
       }
 
+   CFG(TR::Compilation *c, TR::ResolvedMethodSymbol *m, TR::Region &r) :
+      _structureRegion(c->trMemory()->heapMemoryRegion()),
```
Shouldn't this be `_structureRegion(r)`?

andrewcraik commented on compiler/infra/OMRCfg.hpp:
```diff
@@ -312,6 +359,7 @@ class CFG
    TR::CFGNode *_pStart;
    TR::CFGNode *_pEnd;
    TR::Region _structureRegion;
+   TR::Region _internalRegion;
```
This looks wrong - we shouldn't copy the region we should reference it... the _structureRegion looks wrong too. FYI @Leonardo2718.

efferifick replied:
@andrewcraik I thought that the idea was for `OMRCfg` to own the region where edges are stored. This way, once `OMRCfg` is destroyed, the edges are freed.

andrewcraik replied:
Right but if you are defaulting it to heap this will copy the heap which isn't the right thing to do...

andrewcraik replied:
@efferifick never mind - my bad this is good - ignore me
